tldraw/packages/core/src/lib/tools/TLDrawTool/states/CreatingState.tsx

```diff
@@ -96,7 +97,13 @@ export class CreatingState<
     })
     this.tool.previousShape = this.shape
     this.tool.transition('idle')
-    this.app.persist()
+    let tool = this.app.selectedTool.id
+    if (tool === 'pencil' || tool === 'highlighter') {
+      if (this.persistTimeout) { clearTimeout(this.persistTimeout) }
+      this.persistTimeout = setTimeout(this.app.persist, 3000)
```

Comment:
persist should only be called synchronously, otherwise it will break the redo/undo stack

Reply:
Is this to fix the lag issue that when persisting a canvas will cause filesync to sync files? I think we need to think about optimizing `frontend.state/whiteboard-page-idle?`?

Reply:
I suspect it's because persisting a canvas could block the UI.

Reply:
> persist should only be called synchronously, otherwise it will break the redo/undo stack
Maybe we can persist it before undo if there's any unsaved state?

Reply:
> Maybe we can persist it before undo if there's any unsaved state?

I don't think this is going to work either. It will also merge actions that shouldn't be merged. Even if we manage to combine all sequential pencil actions by calling `pause()`, the experience won't be great. I think users would expect to be able to undo each stroke separately.